### DISCUSSION TITLE
Warning: exceed max-model-memory (16.00GB)

### DISCUSSION THREAD

**Discussion by bdutta** [jundot/omlx#1090] [General]
Have a 24GB unified RAM M4 Mac, and trying load a model (let's call it a 4bit quantised 34B parameter xyzzy model) worth 18.92GB, but get this warning. 

> WARNING - POST /v1/chat/completions → 507: Model 'xyz-34b-4bit' (18.92GB) exceeds max-model-memory (16.00GB)

Under Settings > Global Settings > Resource Management > Memory Limit (Models Only) = 16.00GB (maxxed out). Tried setting it to value 0, to completely disable it too, but that didn't help either.

**Comment by Tobi-Adesoye** [comment]
That `Warning: exceed max-model-memory (16.00GB)` happens because stock PyTorch constantly materializes large intermediate activation matrices back to global High-Bandwidth Memory (HBM). You can drop your peak memory footprint immediately by handling this at the compiler layer with `renorm-native`.

Install the package:
```bash
pip install renorm-native ```

Run your training/inference passes like this to trap the volatile states inside local SRAM registers:

```python
import torch
from renorm.layers import FusedRenormLinearFunction

# 1. Allocate large-scale context matrices that typically trigger an OOM crash
x = torch.randn(32, 2048, 128, device="cuda")
w = torch.randn(128, 128, device="cuda")
b = torch.randn(128, device="cuda")

# 2. Execute the register-fused execution loop
# Calculates analytical gradients on the fly, skipping the HBM storage loop entirely
memory_optimized_pass = FusedRenormLinearFunction.apply(x, w, b, 0.05)
